### PR TITLE
[3250][ADD] purchase_order_restriction

### PR DESCRIPTION
**Pull request by rinaldifirdaus**
[3250](https://www.quartile.co/web?db=quartile&token=dMRr78CMkEX6gyy9nTiv#id=3250&action=771&model=project.task&view_type=form&menu_id=505)

### CODE REVIEW

File: purchase_order_restriction/readme/DESCRIPTION.rst
Comment (yostashiro) [inline purchase_order_restriction/readme/DESCRIPTION.rst:1]: Fix the presentation.

image

File: purchase_order_restriction/models/purchase.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+from odoo import fields, models
+
+
+class PurchaseOrder(models.Model):
+    _inherit = "purchase.order"
+
+    restrict_purchase_order = fields.Boolean(
+        "Restrict Order",
+        help="If set as True, purchase order can only be accessed by Purchase Manager",
+    )
```
Comment (yostashiro) [inline purchase_order_restriction/models/purchase.py]:
```suggestion
    is_restricted = fields.Boolean(
        "Restrict Access",
        tracking=True,
        help="If selected, this purchase order can only be accessed by the assigned buyer or purchase managers.",
    )
    
    @api.constrains("is_restricted")
    def _check_is_restricted_access(self):
        for order in self:
            if order.user_id != self.env.user or not self.user_has_groups("purchase.group_purchase_manager"):
                raise ValidationError(_("You do not have the right to change Restric Access setting."))
```

File: purchase_order_restriction/security/security.xml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    <record model="ir.rule" id="purchase_order_restriction_users">
+        <field name="name">View purchase orders for users</field>
+        <field name="model_id" ref="model_purchase_order" />
+        <field name="domain_force">[("restrict_purchase_order","=",False)]</field>
+        <field name="groups" eval="[(4,ref('purchase.group_purchase_user'))]" />
+        <field name="perm_read" eval="True" />
+        <field name="perm_write" eval="True" />
+        <field name="perm_create" eval="True" />
+        <field name="perm_unlink" eval="True" />
+    </record>
```
Comment (yostashiro) [inline purchase_order_restriction/security/security.xml]:
```suggestion
    <record model="ir.rule" id="purchase_order_restriction_users">
        <field name="name">View purchase orders (own or unrestricted)</field>
        <field name="model_id" ref="model_purchase_order" />
        <field name="domain_force">['|',('is_restricted','=',False),('user_id','=',user.id)]</field>
        <field name="groups" eval="[(4,ref('purchase.group_purchase_user'))]" />
    </record>
```

File: purchase_order_restriction/security/security.xml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        <field name="model_id" ref="model_purchase_order" />
+        <field
+            name="domain_force"
+        >["|",("restrict_purchase_order","=",True),("restrict_purchase_order","=",False)]</field>
+        <field name="groups" eval="[(4,ref('purchase.group_purchase_manager'))]" />
+        <field name="perm_read" eval="True" />
+        <field name="perm_write" eval="True" />
+        <field name="perm_create" eval="True" />
+        <field name="perm_unlink" eval="True" />
+    </record>
```
Comment (yostashiro) [inline purchase_order_restriction/security/security.xml]:
```suggestion
    <record model="ir.rule" id="purchase_order_restriction_managers">
        <field name="name">View purchase orders (manager)</field>
        <field name="model_id" ref="model_purchase_order" />
        <field name="domain_force">[1,'=',1]</field>
        <field name="groups" eval="[(4,ref('purchase.group_purchase_manager'))]" />
    </record>
```

File: purchase_order_restriction/views/purchase_views.xml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    </record>
+    <record id="purchase_order_tree" model="ir.ui.view">
+        <field name="name">purchase.order.tree</field>
+        <field name="model">purchase.order</field>
+        <field name="inherit_id" ref="purchase.purchase_order_view_tree" />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='user_id']" position="after">
+                <field
+                    name="restrict_purchase_order"
+                    optional="show"
```
Comment (yostashiro) [inline purchase_order_restriction/views/purchase_views.xml]:
```suggestion
                    optional="hide"
```

File: purchase_order_restriction/views/purchase_views.xml [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    </record>
+    <record id="purchase_order_kpis_tree" model="ir.ui.view">
+        <field name="name">purchase.order.inherit.purchase.order.tree</field>
+        <field name="model">purchase.order</field>
+        <field name="inherit_id" ref="purchase.purchase_order_kpis_tree" />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='user_id']" position="after">
+                <field
+                    name="restrict_purchase_order"
+                    optional="show"
```
Comment (yostashiro) [inline purchase_order_restriction/views/purchase_views.xml]:
```suggestion
                    optional="hide"
```

File: purchase_order_restriction/views/purchase_views.xml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        <field name="model">purchase.order</field>
+        <field name="inherit_id" ref="purchase.purchase_order_form" />
+        <field name="priority">100</field>
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='date_order']" position="after">
+                <field
+                    name="restrict_purchase_order"
+                    groups="purchase.group_purchase_manager"
+                />
+            </xpath>
```
Comment (yostashiro) [inline purchase_order_restriction/views/purchase_views.xml]:
```suggestion
            <xpath expr="//group[@name='other_info']" position="inside">
                <field name="is_restricted" />
            </xpath>
```

File: purchase_order_restriction/views/purchase_views.xml [hunk 7]
```diff
@@ -0,0 +1,45 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="purchase_order_form" model="ir.ui.view">
+        <field name="name">purchase.order.form</field>
+        <field name="model">purchase.order</field>
+        <field name="inherit_id" ref="purchase.purchase_order_form" />
+        <field name="priority">100</field>
```
Comment (yostashiro) [inline purchase_order_restriction/views/purchase_views.xml]: Why this priority?
```suggestion
```

**Review by yostashiro — CHANGES_REQUESTED** [review]
We need to assume that users who create purchase orders are not puerchase managers, and the restriction needs to be applied as necessary right upon creation of the order. Please check and confirm the validity of my suggestions.

**Comment by rinaldifirdaus** [comment]
@yostashiro san, i've added a new commit for this module and add some adjustment from your feedback. please kindly your review.

File: purchase_order_restriction/models/purchase.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+
+class PurchaseOrder(models.Model):
+    _inherit = "purchase.order"
+
+    is_restricted = fields.Boolean(
+        "Restrict Access",
+        tracking=True,
+        help="If selected, this purchase order can only be accessed\
+              by the assigned buyer or purchase managers.",
```
Comment (yostashiro) [inline purchase_order_restriction/models/purchase.py]:
```suggestion
        help="If selected, this purchase order can only be accessed by the assigned "
        "buyer or purchase managers.",
```

File: purchase_order_restriction/views/purchase_views.xml [hunk 9]
```diff
@@ -0,0 +1,41 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="purchase_order_form" model="ir.ui.view">
+        <field name="name">purchase.order.form</field>
+        <field name="model">purchase.order</field>
+        <field name="inherit_id" ref="purchase.purchase_order_form" />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='date_order']" position="before">
```
Comment (yostashiro) [inline purchase_order_restriction/views/purchase_views.xml]: You missed my earlier suggestion. IMO, the field does not need to be always visible as the cases where `is_restricted` needs to be selected should be limited. 

```suggestion
            <xpath expr="//group[@name='other_info']" position="inside">
```

File: purchase_order_restriction/readme/DESCRIPTION.rst [hunk 10]
```diff
@@ -0,0 +1,5 @@
+This module does the following:
+
+* Adds Restrict Access boolean field in purchase order form for Purchase Manager and Purchase Order Owner to mark that the record is restricted
+* Adds new record rule for Purchase Users to have access only into their own record that have Restrict Access value is True
+* Adds Restrict Access to RFQ and Purchase Orders Tree view for Purchase Manager to monitor which records is currently restricted on the list
```
Comment (yostashiro) [inline purchase_order_restriction/readme/DESCRIPTION.rst]: This explanation is too detailed and it does not add much value while it adversely adds to the maintenance cost. We should not be listing every single detail of what's included in the module.

What's important in this documentation are:

- to roughly describes what the module does, and
- to explain the reason/background why we need this module

Our audience are not really the end-users, but the people with knowledge of Odoo development, who would just grasp the implementation details by reading the code.

```suggestion
This module adds Ristrict Access boolean field in the purchase order model, to limit the
access to the purchase order to 1) the buyer and 2) purchase managers.

The module can be useful if there are orders whose conditions should not be visible to
wide range of buyers (e.g. contract rates of the temp staff).
```

**Review by yostashiro — CHANGES_REQUESTED** [review]

**Review by yostashiro — APPROVED** [review]

**Review by AungKoKoLin1997 — APPROVED** [review]

**Comment by AungKoKoLin1997** [comment]
@rinaldifirdaus This module is the same feature with https://github.com/OCA/purchase-workflow/tree/15.0/purchase_security, isn't it? Did you check in OCA before you implemented this module or I miss something?

**Comment by yostashiro** [comment]
@rinaldifirdaus Add the translation pls.

**Comment by yostashiro** [comment]
Superseded by https://github.com/qrtl/axls-oca/pull/54.

**Pull request closed without merging**